cmd/dptp-controller-manager/main.go

```diff
@@ -236,24 +235,33 @@ func main() {
 	ctx := controllerruntime.SetupSignalHandler()
 	ctx, cancel := context.WithCancel(ctx)
 
-	kubeconfigChangedCallBack := func(e fsnotify.Event) {
-		logrus.WithField("event", e.String()).Info("Kubeconfig changed, exiting to get restarted by Kubelet and pick up the changes")
+	kubeconfigChangedCallBack := func() {
+		logrus.Info("Kubeconfig changed, exiting to get restarted by Kubelet and pick up the changes")
 		cancel()
 	}
 
-	kubeconfigs, err := util.LoadKubeConfigs(opts.kubeconfig, opts.kubeconfigDir, kubeconfigChangedCallBack)
+	kubeconfigs, err := opts.kubernetesOptions.LoadClusterConfigs(kubeconfigChangedCallBack)
 	if err != nil {
 		logrus.WithError(err).Fatal("failed to load kubeconfigs")
 	}
+
+	inClusterConfig, hasInClusterConfig := kubeconfigs[kube.InClusterContext]
+	delete(kubeconfigs, kube.InClusterContext)
+	delete(kubeconfigs, kube.DefaultClusterAlias)
+
 	if _, hasAppCi := kubeconfigs[appCIContextName]; !hasAppCi {
-		kubeconfigs[appCIContextName], err = rest.InClusterConfig()
-		if err != nil {
-			logrus.WithError(err).Fatalf("--kubeconfig had no context for '%s' and loading InClusterConfig failed", appCIContextName)
-		}
-		logrus.Infof("Loaded %q context from in-cluster config", appCIContextName)
-		if err := util.WatchFiles([]string{"/var/run/secrets/kubernetes.io/serviceaccount/token"}, kubeconfigChangedCallBack); err != nil {
-			logrus.WithError(err).Fatal("faild to watch in-cluster token")
+		if !hasInClusterConfig {
+			logrus.WithError(err).Fatalf("had no context for '%s' and loading InClusterConfig failed", appCIContextName)
+		} else {
+			// There is a corner case that the inClusterConfig is the current context of a loaded kubeconfig file
+			// In dev, it has to be app.ci's public URL or in a pod, it ${KUBERNETES_SERVICE_HOST}:${KUBERNETES_SERVICE_PORT}
+			// https://github.com/kubernetes/test-infra/blob/5bce0be55da8a04f62c77d2292a81739491544df/prow/kube/config.go#L56-L62
+			inClusterConfigHost := fmt.Sprintf("%s:%s", os.Getenv("KUBERNETES_SERVICE_HOST"), os.Getenv("KUBERNETES_SERVICE_PORT"))
```

Comment:
But why? This code shouldn't reimplement part of the client-go logic (and break if client-go should be changed). So far, we trusted that a context named `app.ci` is for app.ci and that the in-cluster config is for that as well. What is the reason not to trust that anymore?

Reply:
For example, if we accidentally removed this line

https://github.com/openshift/ci-tools/blob/cc37aa0f41d5e5d71739d87123433e6d34ec7d9c/hack/run-dptp-controller-manager.sh#L11

then we would use another cluster as `app.ci`.

I just want to be safer.

Reply:
Previously, we load `inCluster` without the logic from upstream and now the upstream logic could give me a wrong result.

Reply:
Why would in the example give a wrong result?

Reply:
It could come from `currentContext` of the last kubeconfig file it loads.

https://github.com/kubernetes/test-infra/blob/c3f3e307169f26b3790a132060c4bb2f3c76a66a/prow/kube/config.go#L58

and

https://github.com/kubernetes/test-infra/blob/c3f3e307169f26b3790a132060c4bb2f3c76a66a/prow/kube/config.go#L121-L135

Reply:
Ok, I am just uneasy about relying on implementation details here. The thing needs to be prefixed with `https://` in the in-cluster case and joined with net.JoinHostPorT: https://github.com/openshift/release-controller/blob/efd4b6d1c92ff73804e397524de974d160a8dc31/vendor/k8s.io/client-go/rest/config.go#L508

Reply:
Updated.